subsys/bluetooth/services/mds.c

```diff
@@ -0,0 +1,544 @@
+#include <zephyr/bluetooth/gatt.h>
+
+#include <bluetooth/services/mds.h>
+
+#include <zephyr/net/buf.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/sys/__assert.h>
+
+#include <memfault/config.h>
+#include <memfault/core/platform/device_info.h>
+#include <memfault/core/data_packetizer.h>
+
+LOG_MODULE_REGISTER(mds, CONFIG_BT_MDS_LOG_LEVEL);
+
+#ifndef CONFIG_BT_MDS_PERM_RW
+#define CONFIG_BT_MDS_PERM_RW 0
+#endif
+#ifndef CONFIG_BT_MDS_PERM_RW_ENCRYPT
+#define CONFIG_BT_MDS_PERM_RW_ENCRYPT 0
+#endif
+#ifndef CONFIG_BT_MDS_PERM_RW_AUTHEN
+#define CONFIG_BT_MDS_PERM_RW_AUTHEN 0
+#endif
+
+#define MDS_GATT_PERM_READ (                                                \
+	CONFIG_BT_MDS_PERM_RW_AUTHEN ?                                      \
+		BT_GATT_PERM_READ_AUTHEN : (CONFIG_BT_MDS_PERM_RW_ENCRYPT ? \
+			BT_GATT_PERM_READ_ENCRYPT : BT_GATT_PERM_READ)      \
+	)
+
+#define MDS_GATT_PERM_WRITE (                                                \
+	CONFIG_BT_MDS_PERM_RW_AUTHEN ?                                       \
+		BT_GATT_PERM_WRITE_AUTHEN : (CONFIG_BT_MDS_PERM_RW_ENCRYPT ? \
+			BT_GATT_PERM_WRITE_ENCRYPT : BT_GATT_PERM_WRITE)     \
+	)
+
+#define MDS_MAX_URI_LENGTH CONFIG_BT_MDS_MAX_URI_LENGTH
+
+#define MDS_URI_BASE \
+	MEMFAULT_HTTP_APIS_DEFAULT_SCHEME "://" MEMFAULT_HTTP_CHUNKS_API_HOST "/api/v0/chunks/"
+
+#define MDS_AUTH_KEY "Memfault-Project-Key:" CONFIG_MEMFAULT_NCS_PROJECT_KEY
+
+#define DATA_POLL_INTERVAL CONFIG_BT_MDS_DATA_POLL_INTERVAL
+
+/* Memfault chunk number maximum value. Chunk number should overlaps after reaching this value. */
+#define MDS_CHUNK_NUMBER_MAX_VALUE BIT_MASK(5)
+
+#define MAX_PIPELINE CONFIG_BT_MDS_PIPELINE_COUNT
+
+#define STREAM_ENABLED BIT(0)
+
+/* Application error code defined by the MDS.
+ * According to BLE Core v5.3 Vol 3, Part F 3.4.1.
+ */
+enum mds_att_error {
+	MDS_ATT_ERROR_CLIENT_ALREADY_SUBSCRIBED = 0x80,
+	MDS_ATT_ERROR_CLIENT_NOT_SUBSCRIBED     = 0x81
+};
+
+enum data_export_mode {
+	DATA_EXPORT_MODE_STREAMING_DISABLE = 0x00,
+	DATA_EXPORT_MODE_STREAMING_ENABLE  = 0x01
+};
+
+enum mds_read_char {
+	MDS_READ_CHAR_SUPPORTED_FEATURES,
+	MDS_READ_CHAR_DEVICE_IDENTIFIER,
+	MDS_READ_CHAR_DATA_URI,
+	MDS_READ_CHAR_AUTHORIZATION
+};
+struct mds {
+	const struct bt_mds_cb *cb;
+	struct bt_conn *conn;
+	atomic_t send_cnt;
+	atomic_t stream_state;
+	uint8_t chunk_number;
+};
+
+struct mds_data_export_nfy {
+	uint8_t chunk_number:5;
+	uint8_t rfu:3;
+	uint8_t data[];
+};
+
+struct mds_subscription {
+	bool subscribed;
+	const struct bt_conn *conn;
+	const struct bt_gatt_attr *attr;
+};
+
+static struct mds mds_instance = {
+	.send_cnt = ATOMIC_INIT(MAX_PIPELINE)
+};
+
+static void mds_work_handler(struct k_work *work);
+
+K_WORK_DELAYABLE_DEFINE(mds_work, mds_work_handler);
+
+/* System workqueue thread cannot be preempted to ensure proper service operations.
+ * Make sure here that system workqueue is a cooperative thread.
+ */
+BUILD_ASSERT(CONFIG_SYSTEM_WORKQUEUE_PRIORITY < 0);
+
+
```

Comment:
Unnecessary blank line.

Reply:
I wanted to add two blank lines before first function definition.

Reply:
OK, I see.